### PR TITLE
Overflow86

### LINKED ISSUE #86: steady time overflow issue
As reported by @serge-nikulin, this test code will eventually produce unexpected output on Windows (I do not think it affects Linux/macOS):

```c++
int main()
{
  rcutils_time_point_value_t last_time_ns;
  rcutils_time_point_value_t this_time_ns;
  rcutils_steady_time_now(&last_time_ns);
  for (uint64_t count = 0ULL; ; ++count)
  {
    rcutils_steady_time_now(&this_time_ns);
    uint64_t delta_ns = this_time_ns - last_time_ns;
    if (delta_ns > 100000000ULL) {
      std::cout << "****** Unexpected delta: " << delta_ns << "*********\n";
    }
    if ((count % 1000000000ULL) == 0ULL) {
      std::ostringstream ss;
      ss.imbue(std::locale(""));
      ss << count;
      std::cout << ss.str() << "\n";
    }
    last_time_ns = this_time_ns;
  }
  return 0;
}
```

Something like this (after about 30 minutes or so):

```
41,000,000,000
42,000,000,000
43,000,000,000
44,000,000,000
45,000,000,000
****** Unexpected delta: 18446738826622249502*********
46,000,000,000
47,000,000,000
48,000,000,000
```

The theory is that this is due to an overflow of an intermediate calculation which uses the performance ticks on a computer with a GHz speed processor.

@serge-nikulin said:

> The culprit is `rcutils_steady_time_now`: multiplication by `10^9` periodically overflows `uint64_t` (max ~`10^19`).
With GHZ frequencies (`10^9`) it should be a matter of time to get an overflow condition:

```
  // Convert to nanoseconds before converting from ticks to avoid precision loss.
  rcutils_time_point_value_t intermediate = RCUTILS_S_TO_NS(performance_count.QuadPart);
  *now = intermediate / cpu_frequency.QuadPart;
```
> I already have brushed this bug in https://github.com/ros2/rcutils/issues/79#issuecomment-362406669.

### PR DESCRIPTION
**Pull request by serge-nikulin**
closes #86 

@wjwwood, I am based on ROS2 master but somehow I'm picking up signed time changes.
The only intended change is in `rcutils_steady_time_now` function. This change should be compatible with both signed and unsigned time.

```C
rcutils_ret_t
rcutils_steady_time_now(rcutils_time_point_value_t * now)
{
  RCUTILS_CHECK_ARGUMENT_FOR_NULL(
    now, RCUTILS_RET_INVALID_ARGUMENT, rcutils_get_default_allocator());
  LARGE_INTEGER cpu_frequency, performance_count;
  // These should not ever fail since XP is already end of life:
  // From https://msdn.microsoft.com/en-us/library/windows/desktop/ms644905(v=vs.85).aspx and
  //      https://msdn.microsoft.com/en-us/library/windows/desktop/ms644904(v=vs.85).aspx:
  // "On systems that run Windows XP or later, the function will always succeed and will
  //  thus never return zero."
  QueryPerformanceFrequency(&cpu_frequency);
  QueryPerformanceCounter(&performance_count);
  // Calculate nanoseconds and seconds separately because
  // otherwise overflow can happen in intermediate calculations
  // This conversion will overflow if the PC runs >292 years non-stop
  const rcutils_time_point_value_t whole_seconds =
    performance_count.QuadPart / cpu_frequency.QuadPart;
  const rcutils_time_point_value_t remainder_count =
    performance_count.QuadPart % cpu_frequency.QuadPart;
  const rcutils_time_point_value_t remainder_ns =
    RCUTILS_S_TO_NS(remainder_count) / cpu_frequency.QuadPart;
  const rcutils_time_point_value_t total_ns =
    RCUTILS_S_TO_NS(whole_seconds) + remainder_ns;
  *now = total_ns;
  return RCUTILS_RET_OK;
}
```

### CODE REVIEW

**Comment by mikaelarguedas** [comment]
I suspect that your local master branch is not up to date.
Assuming that the ros2/rcutils repository is your "origin" remote (you can check your remote names and URLs by running `git remote -v`), you should be able to rebase on to of the current master by doing:
```
git fetch origin
git rebase origin/master
git push -f myfork overflow86
```
Hope this helps

**Comment by serge-nikulin** [comment]
@mikaelarguedas, thanks!

It looks good now.

File: src/time_win32.c [hunk 1]
```diff
@@ -60,11 +60,18 @@ rcutils_steady_time_now(rcutils_time_point_value_t * now)
...
-  // Convert to nanoseconds before converting from ticks to avoid precision loss.
-  uint64_t intermediate = RCUTILS_S_TO_NS(performance_count.QuadPart);
-  intermediate /= cpu_frequency.QuadPart;
+  // Calculate nanoseconds and seconds separately because
+  // otherwise overflow can happen in intermediate calculations
   // This conversion will overflow if the PC runs >292 years non-stop
-  *now = (rcutils_time_point_value_t)intermediate;
+  const rcutils_time_point_value_t whole_seconds =
+    performance_count.QuadPart / cpu_frequency.QuadPart;
+  const rcutils_time_point_value_t reminder_count =
```
Comment (wjwwood) [inline src/time_win32.c]: Do you mean "remainder" here? Reminder doesn't really make sense to me.

Comment (serge-nikulin) [reply src/time_win32.c]: Fixed!

**Comment by wjwwood** [comment]
Thanks! Here's CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/3942/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/1060/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/3272/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/4041/)

**Comment by serge-nikulin** [comment]
@mikaelarguedas, according to https://github.com/ros2/rcutils/issues/79#issuecomment-362104626, the build is OK, right?

Thanks!

**Comment by mikaelarguedas** [comment]
Yep! failure not due to this PR, I'll leave it to @wjwwood's discretion to merge as I didnt review or tested this.

**Comment by wjwwood** [comment]
Yeah, those composition tests are still known to be flakey from time to time.

Thanks for the contribution!

**Pull request merged**